dll/cpl/input/settings_page.c

```diff
@@ -11,192 +11,341 @@
+    if (iDefault == (INT)SendMessageW(hwndCombo, CB_GETCOUNT, 0, 0))
+        SendMessageW(hwndCombo, CB_SETCURSEL, iDefault - 1, 0);
+    else
+        SendMessageW(hwndCombo, CB_SETCURSEL, iDefault, 0);
+}
 
-                        PatBlt(hdc, 0, 0, width, height, BLACKNESS);
+static VOID
+SetControlsState(HWND hwndDlg)
+{
+    HWND hwndList = GetDlgItem(hwndDlg, IDC_KEYLAYOUT_LIST);
+    HWND hwndCombo = GetDlgItem(hwndDlg, IDC_DEFAULT_LANGUAGE);
+    BOOL bIsLeaf, bCanRemove, bCanProp;
+    HTREEITEM hSelected = TreeView_GetSelection(hwndList);
+    TV_ITEM item = { TVIF_PARAM };
+    item.hItem = hSelected;
 
-                        SelectObject(hdc, hBmpOld);
+    bIsLeaf = (hSelected && TreeView_GetItem(hwndList, &item) && HIWORD(item.lParam));
 
-                        IconInfo.hbmColor = hBitmap;
-                        IconInfo.hbmMask = hBmpNew;
-                        IconInfo.fIcon = TRUE;
+    bCanRemove = bIsLeaf && (s_nLeavesCount > 1);
+    bCanProp = bIsLeaf;
 
-                        hIcon = CreateIconIndirect(&IconInfo);
+    EnableWindow(GetDlgItem(hwndDlg, IDC_REMOVE_BUTTON), bCanRemove);
+    EnableWindow(GetDlgItem(hwndDlg, IDC_PROP_BUTTON), bCanProp);
 
-                        DeleteObject(hBmpOld);
-                    }
+    InitDefaultLangComboBox(hwndCombo);
+}
 
-                    DeleteObject(hFont);
-                }
-            }
+HTREEITEM FindLanguageInList(HWND hwndList, LPCTSTR pszLangName)
+{
+    TV_ITEM item;
+    TCHAR szText[128];
+    HTREEITEM hItem;
 
-            DeleteObject(hBmpNew);
-        }
+    hItem = TreeView_GetRoot(hwndList);
+    while (hItem)
+    {
+        szText[0] = 0;
+        item.mask       = TVIF_TEXT | TVIF_HANDLE;
+        item.pszText    = szText;
+        item.cchTextMax = _countof(szText);
+        item.hItem      = hItem;
+        TreeView_GetItem(hwndList, &item);
+        if (lstrcmpi(szText, pszLangName) == 0)
+            return hItem;
+
+        hItem = TreeView_GetNextSibling(hwndList, hItem);
     }
 
-    DeleteDC(hdc);
-    DeleteObject(hBitmap);
-
-    return hIcon;
+    return NULL;
 }
 
-
 static VOID
-SetControlsState(HWND hwndDlg)
+AddToInputListView(HWND hwndList, INPUT_LIST_NODE *pInputNode)
 {
-    HWND hwndList = GetDlgItem(hwndDlg, IDC_KEYLAYOUT_LIST);
-    INT iSelected = ListView_GetNextItem(hwndList, -1, LVNI_SELECTED);
-    INT nCount = ListView_GetItemCount(hwndList);
-    BOOL bCanRemove = (iSelected != -1) && (nCount >= 2);
-    BOOL bCanDefault = (iSelected != -1) && (nCount >= 2);
-    BOOL bCanProp = (iSelected != -1);
-
-    LV_ITEM item = { LVIF_PARAM, iSelected };
-    if (ListView_GetItem(hwndList, &item))
+    INT LangImageIndex = -1, KeyboardImageIndex = -1, ImeImageIndex = -1;
+    HICON hLangIcon, hKeyboardIcon, hInputMethodIcon;
+    TV_ITEM item;
+    TV_INSERTSTRUCT insert;
+    HIMAGELIST hImageList = TreeView_GetImageList(hwndList, TVSIL_NORMAL);
+    TCHAR szKeyboard[] = TEXT("Keyboard");
```

Comment:
Should this need to be localized? or not?
Also it seems you could just have used something like:
```suggestion
    PCWSTR pszKeyboard = L"Keyboard";
```
or even directly do `item.pszText = "Keyboard";` below :thinking:

Reply:
Used `LoadStringW`. Done.